docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,365 @@
+
+As introduced above, we propose adding a new gtscript `region()` specifier that specializes the stencil on a region of the space when a non-zero intersection with the global region occurs.
+The idea behind using the region objects is that they can incorporate both the iteration space a process sees and the global partitioning that an application must handle, all while hiding the complexity at the stencil level for model developers.
+In this proposal, the application remains in control of the partitioning data so **the partitioner is entirely transparent to both GT4Py and its backend.**
+
+
+Specification Options
++++++++++++++++++++++
+
+An issue we faced when designing the iteration space featiure is that the stages (`with interval()`) are, by the execution model specification, free to be re-ordered by the runtime.
+We propose adding the `region()` to **either** (1) the multistage, in which case the execution order is specified by GridTool's execution model **or** (2) within a stage.
+We plan to implement these in order, first finishing an implementation of (1) before attempting the second.
+If we agree to try (2) the example would look like:
+
+.. code-block:: python
+
+    @gtscript.stencil(backend=backend_type)
+    def stencil(uc: FIELD_TYPE, vc: FIELD_TYPE, cosa: FIELD_TYPE, rsina: FIELD_TYPE, ut: FIELD_TYPE, ub: FIELD_TYPE, dt5: float, dt4: float):
+        with computation(PARALLEL), interval(...):
+            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
+            with region(WestEdge(0, 1), EastEdge(0, 1)):
+                ub = dt5 * (ut[0, -1, 0] + ut)
+            with region(SouthEdge(0, 1), NorthEdge(0, 1)):
+                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
+This creates a switch-like stage where the execution at the point choses one of these options.
+The first line(s), or the lines ouside a `with` block, specify the default operation, and specializations are inside the blocks.
+An ambiguity arises when the same point is represented in multiple specializations, but that is easily cleared up by specifying that the **last** specialization is the one that takes precedence.
+The reason for choosing this option is that performance (especially in the CUDA backend) is retained, even without multistage fusion, since the backend does not see multiple multistages.
+However, this has the drawback that additional parsing is needed.
+
+
+Relation to Global Iteration Space
+++++++++++++++++++++++++++++++++++
+
+To explain how the specification covers the domain decomposition case, consider part of the example above with a domain of size N by N by K points, decomposed by three processes using vertical strips of the domain:
+
+.. math:: \mathrm{proc}\, 0: [0,n) \times [0,N) \times [0,K),\quad \mathrm{proc}\, 1: [n,2n) \times [0,N) \times [0,K),\quad \mathrm{proc}\, 2: [2n,N) \times [0,N) \times [0,K).
+
+.. code-block:: python
+
+    with region(WestEdge(0, 1), EastEdge(0, 1)):
```

Comment:
Can this be defined as "the set-union" of the points of the west edge and the east edge"?
In this case a `with region(WestEdge(0, 1), SouthEdge(0, 1)):` would have no ambiguity since the intersection is logically evaluated once.

Reply:
The way it works in practice is that a new `ComputationBlock` is generated for each argument to `region()`, which is not quite equivalent to the set union, in the case where fields are incremented (since they are computed twice).

We had floated the idea of adding set operations on the region objects. Would syntax like `union(WestEdge(0, 1), SouthEdge(0, 1))` be helpful to users?

After starting the implementation, my concern is adding complexity: the regions (each argument to `region()`) are currently limited to connected convex regions, and allowing unions would break this imposed requirement.